### PR TITLE
NODDIx 2 fibers crossing

### PR DESCRIPTION
**Pull request by ShreyasFadnavis**
- Create Example for the NODDIx model

### CODE REVIEW

**Comment by ShreyasFadnavis** [comment]
[GSoC 2018 - PSF - DIPY report.pdf](https://github.com/nipy/dipy/files/2280920/GSoC.2018.-.PSF.-.DIPY.report.pdf)

File: dipy/reconst/NODDIx.py [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+    def __init__(self, gtab, params, fit_method='MIX'):
+        # The maximum number of generations, genetic algorithm 1000 default, 1
+        self.maxiter = 100
+        # Tolerance for termination, nonlinear least square 1e-8 default, 1e-3
+        self.xtol = 1e-8
+        self.gtab = gtab
+        self.big_delta = gtab.big_delta
+        self.small_delta = gtab.small_delta
+        self.gamma = gamma
+        self.G = params[:, 3] / 10 ** 6  # gradient strength (Tesla/micrometer)
```
Comment (skoudoro) [inline dipy/reconst/NODDIx.py]: is it possible to have a more explicit variable name?

File: dipy/reconst/NODDIx.py [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+
+        bounds = ([0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01,
+                   0.01], [0.9, 0.9, 0.9, 0.9, 0.9, 0.99, np.pi, np.pi, 0.99,
+                           np.pi, np.pi])
+        res = least_squares(self.nlls_cost, x_f, xtol=self.xtol, args=(data,))
+        result = res.x
+        return result
+
+    def stoc_search_cost(self, x, signal):
+        """
```
Comment (skoudoro) [inline dipy/reconst/NODDIx.py]: the doc string miss `Parameter`

File: dipy/reconst/NODDIx.py [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+        --------
+        cost function for differential evolution algorithm:
+        .. math::
+            (signal -  S)^T(signal -  S)
+        """
+        phi = self.Phi(x)
+        return self.differential_evol_cost(phi, signal)
+
+    def differential_evol_cost(self, phi, signal):
+
```
Comment (skoudoro) [inline dipy/reconst/NODDIx.py]: same as above. No documentation

File: dipy/reconst/NODDIx.py [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+        return np.dot((signal - yhat).T, signal - yhat)
+
+    def cvx_fit(self, signal, phi):
+        """
+        Linear parameters fit using cvx
+        Parameters
+        ----------
+        phi : array
+            phi.shape = number of data points x 4
+            signal : array
```
Comment (skoudoro) [inline dipy/reconst/NODDIx.py]: Can you remove a tab for the alignment

File: dipy/reconst/NODDIx.py [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+        obj = cvx.Minimize(cvx.sum_entries(cvx.square(phi * f - signal)))
+
+        # Form and solve problem.
+        prob = cvx.Problem(obj, constraints)
+        # Returns the optimal value
+        prob.solve()
+        return np.array(f.value)
+
+    def nlls_cost(self, x_f, signal):
+        """
```
Comment (skoudoro) [inline dipy/reconst/NODDIx.py]: doc string missing "signal"

File: dipy/reconst/noddi_speed.pyx [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+def legendre_gauss_integral(double[:] x_vec, cnp.npy_intp n):
+    """
+    Computes legendre gaussian integrals up to the order specified and the
+    derivatives if requested.
+
+    The integral takes the following form, in Mathematica syntax:
+    L[x, n] = Integrate[Exp[-x \mu^2] Legendre[2*n, \mu], {\mu, -1, 1}]
+    L[x, n] = Integrate[Exp[-x \mu^2] (-\mu^2) Legendre[2*n, \mu], {\mu, -1, 1}]
+
+    INPUTS:
```
Comment (skoudoro) [inline dipy/reconst/noddi_speed.pyx]: Parameters instead

File: dipy/reconst/noddi_speed.pyx [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        x should be a column vector of positive numbers, specifying the
+        parameters of the gaussian
+        n should be a non-negative integer, such that 2n specifies the maximum
+        order of legendre polynomial
+        The maximum value for n is 6.
+
+    OUTPUTS:
```
Comment (skoudoro) [inline dipy/reconst/noddi_speed.pyx]: Returns instead

File: dipy/reconst/noddi_speed.pyx [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    return L
+
+def watson_sh_coeff(double k):
+    """
+    Computes the spherical harmonic (SH) coefficients of the Watson's
+    distribution with the concentration parameter k (kappa) up to the 12th
+    order and the derivatives if requested.
+
+    Truncating at the 12th order gives good approximation for kappa up to 64.
+    INPUTS:
```
Comment (skoudoro) [inline dipy/reconst/noddi_speed.pyx]: same as above

File: dipy/reconst/noddi_speed.pyx [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+        k should be an array of positive numbers, specifying a set of
+        concentration parameters for the Watson's distribution.
+
+    OUTPUTS:
```
Comment (skoudoro) [inline dipy/reconst/noddi_speed.pyx]: same as above

File: dipy/sims/sim_noddix.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+"""
+Lets contruct the signal now
+"""
+x_f_sig = [volfrac_ic1, volfrac_ic2, volfrac_ec1, volfrac_ec2, volfrac_csf,
+           OD1, theta1, phi1, OD2, theta2, phi2]
+f = x_f_sig[0:5]
+
+phi = noddix_model.Phi2(x_f_sig)
+reconst_signal = np.dot(phi, f)
```
Comment (skoudoro) [inline dipy/sims/sim_noddix.py]: Can you create a setup function for the bloc above?

File: dipy/reconst/multi_voxel.py [hunk 11]
```diff
@@ -30,6 +30,7 @@ def new_fit(self, data, mask=None):
         fit_array = np.empty(data.shape[:-1], dtype=object)
         for ijk in ndindex(data.shape[:-1]):
             if mask[ijk]:
+                print(ijk)
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply dipy/reconst/multi_voxel.py]: Yes

File: dipy/data/__init__.py [hunk 12]
```diff
@@ -288,6 +288,10 @@ def get_data(name='small_64D'):
         return pjoin(DATA_DIR, 'cb_2.npz')
     if name == "t1_coronal_slice":
         return pjoin(DATA_DIR, 't1_coronal_slice.npy')
+    if name == "small_NODDIx_data":
+        fimg = pjoin(DATA_DIR, 'small_NODDIx_HCP.hdr')
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply dipy/data/__init__.py]: Added

File: dipy/reconst/NODDIx.py [hunk 13]
```diff
@@ -0,0 +1,699 @@
...
+           White Matter Fibers from diffusion MRI." Scientific reports 6
+           (2016).
+
+    Notes
+    -----
+    The implementation of NODDIx may require CVXPY (http://www.cvxpy.org/).
+    """
+
+    def __init__(self, gtab, params, fit_method='MIX'):
+        # The maximum number of generations, genetic algorithm 1000 default, 1
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply dipy/reconst/NODDIx.py]: Cool!

**Pull request closed without merging**